django/db/models/fields/mixins.py

```diff
@@ -24,3 +28,37 @@ def set_cached_value(self, instance, value):
 
     def delete_cached_value(self, instance):
         del instance._state.fields_cache[self.get_cache_name()]
+
+
+def _is_immutable(value):
+    return value is copy(value)
+
+
+class CheckFieldDefaultMixin:
+    """Check the default value of a field to make sure it's not a mutable object."""
+    _default_hint = ('<valid default>', '<invalid default>')
+
+    def _check_default(self):
+        if self.has_default() and not (
+            _is_immutable(self.default) or callable(self.default)
+        ):
+            return [
+                checks.Warning(
+                    "%s default should not be a mutable object so "
+                    "that it's not shared between all field instances." % (
+                        self.__class__.__name__,
+                    ),
+                    hint=(
+                        'Use a callable instead, e.g., use `%s` instead of '
+                        '`%s`.' % self._default_hint
+                    ),
+                    obj=self,
+                )
+            ]
+        else:
+            return []
```

Comment:
@laymonage I don't think that we need any changes in this mixin, previous behavior that checked if `default` is a callable was correct, IMO. We shouldn't allow for non-callable defaults, that's the safest option. If you agree I can move `CheckFieldDefaultMixin` from the `contrib.postgres.fields.mixin` to `models.fields.mixin` in a separate commit.

Reply:
I don't really mind, but I was just following [this part](https://docs.djangoproject.com/en/2.2/ref/models/fields/#default) of the docs that says:

>`Field.default`
> The default value for the field. **This can be a value or a callable object.** If callable it will be called every time a new object is created.

Reply:
I think we can leave a previous behavior because keeping primitives in a `JSONField` is an edge case and providing immutable/primitive defaults is even more niche. Nobody complained about this check that we added in Django 2.1. My understanding is that all users would like to have a `dict` as a default.

Reply:
Yeah, that makes sense. Go ahead :+1:

Reply:
#11932